### PR TITLE
Disable 'shell' environment by default

### PR DESCRIPTION
**Pull request by sbc100**
Including support for shell environment such as d8 and jsc seems like a
very unusual thing to want to do.  This change disables shell support by
default and requires folks that want shell support to specify it with
`-sENVIRONMENT`.

### CODE REVIEW

File: ChangeLog.md [hunk 1]
```diff
@@ -20,6 +20,10 @@ See docs/process.md for more on how version tagging works.
 
 2.0.25
 ------
+- Support for the 'shell' environment by now disabled default.  Running under
```
Comment (kripken) [inline ChangeLog.md:23]:
```suggestion
- Support for the 'shell' environment is now disabled by default.  Running under
```

File: ChangeLog.md [hunk 2]
```diff
@@ -20,6 +20,10 @@ See docs/process.md for more on how version tagging works.
...
+  `d8` or `jsc` is not something that most emscripten users ever want to do.
```
Comment (kripken) [inline ChangeLog.md:24]:
```suggestion
  `d8`, `js`, or `jsc` is not something that most emscripten users ever want to do,
  and including the code for that has been unnecessary for them.
```

**Review by kripken — COMMENTED** [review]
This seems like a better tradeoff than before, I agree.

How about adding code to `shell.js` that in assertions builds, will mention the shell issue? That is, still checking if the environment is a shell there, and then showing a useful error.

**Comment by sbc100** [comment]
> This seems like a better tradeoff than before, I agree.
> 
> How about adding code to `shell.js` that in assertions builds, will mention the shell issue? That is, still checking if the environment is a shell there, and then showing a useful error.

Good idea... I even thought of that while working on this, but then forgot to do it.  Great minds.

**Comment by sbc100** [comment]
Added some extra assertions.

File: src/shell.js [hunk 3]
```diff
@@ -92,12 +92,23 @@ var ENVIRONMENT_IS_WORKER = {{{ ENVIRONMENT === 'worker' }}};
...
 var ENVIRONMENT_IS_NODE = typeof process === 'object' && typeof process.versions === 'object' && typeof process.versions.node === 'string';
 var ENVIRONMENT_IS_SHELL = !ENVIRONMENT_IS_WEB && !ENVIRONMENT_IS_NODE && !ENVIRONMENT_IS_WORKER;
+
+#if ASSERTIONS
+#if !ENVIRONMENT_MAY_BE_SHELL
+assert(!ENVIRONMENT_IS_SHELL, "shell environment detected but not enabled at build time.  Add 'shell' to `-s ENVIRONMENT` to enable shell support.");
+#endif
+#if !ENVIRONMENT_MAY_BE_NODE
+assert(!ENVIRONMENT_IS_NODE, "node environment detected but not enabled at build time.  Add 'node' to `-s ENVIRONMENT` to enable shell support.");
+#endif
```
Comment (kripken) [inline src/shell.js:109]: may as well add "web" and "worker" here as well?

File: src/shell.js [hunk 4]
```diff
@@ -92,12 +92,23 @@ var ENVIRONMENT_IS_WORKER = {{{ ENVIRONMENT === 'worker' }}};
...
 // also a web environment.
 var ENVIRONMENT_IS_NODE = typeof process === 'object' && typeof process.versions === 'object' && typeof process.versions.node === 'string';
 var ENVIRONMENT_IS_SHELL = !ENVIRONMENT_IS_WEB && !ENVIRONMENT_IS_NODE && !ENVIRONMENT_IS_WORKER;
+
+#if ASSERTIONS
+#if !ENVIRONMENT_MAY_BE_SHELL
+assert(!ENVIRONMENT_IS_SHELL, "shell environment detected but not enabled at build time.  Add 'shell' to `-s ENVIRONMENT` to enable shell support.");
+#endif
+#if !ENVIRONMENT_MAY_BE_NODE
+assert(!ENVIRONMENT_IS_NODE, "node environment detected but not enabled at build time.  Add 'node' to `-s ENVIRONMENT` to enable shell support.");
```
Comment (kripken) [inline src/shell.js:108]:
```suggestion
assert(!ENVIRONMENT_IS_NODE, "node environment detected but not enabled at build time.  Add 'node' to `-s ENVIRONMENT` to enable node support.");
```

**Review by kripken — APPROVED** [review]

**Comment by kripken** [comment]
Please also add a test for showing that assertion. That could avoid us losing the helpful message in a future refactor.

**Comment by kripken** [comment]
This may have broken the emscripten roller,

https://chromium-review.googlesource.com/c/emscripten-releases/+/2993293

There is another PR of yours in that roll, but this one seems more likely?

**Comment by sbc100** [comment]
Its not clear how this PR could be to blame.   The roll you sent is interesting because its that same test failing on both mac and windows (wasm0.test_pthread_exit_process).. but I suspect that test might be racy.. perhaps some recent change made it more racy.

It we look at the emscripten roller it has 4 failures now: https://autoroll.skia.org/r/emscripten-emscripten-releases

- https://chromium-review.googlesource.com/c/2993946 - fails just with " Windows 8.1 SDK was not found " 
- https://chromium-review.googlesource.com/c/2993138 - fails just with " Windows 8.1 SDK was not found " 
- https://chromium-review.googlesource.com/c/2993699 - fails just with " Windows 8.1 SDK was not found " 
- https://chromium-review.googlesource.com/c/2993293 - fails on mac and windows with test_pthread_exit_process

What I don't understand is why the `Windows 8.1 SDK was not found` issue would be specific to the emscripten rollers.. it should be a universal failure across all rollers, no?

**Comment by sbc100** [comment]
Wait.. maybe it could be #14536

**Pull request merged**